### PR TITLE
Enrich join API

### PR DESCRIPTION
**Pull request by vitorenesduarte**

### CODE REVIEW

File: src/partisan_static_peer_service_manager.erl [hunk 1]
```diff
@@ -164,21 +161,14 @@ handle_call({reserve, _Tag}, _From, State) ->
...
 
-handle_call({join, {Name, _, _}=Node},
+handle_call({join, {_, _, _}=Node},
             _From,
-            #state{pending=Pending0, connections=Connections0}=State) ->
-    %% Attempt to join via disterl for control messages during testing.
-    _ = net_kernel:connect(Name),
-
-    %% Add to list of pending connections.
-    Pending = [Node|Pending0],
```
Comment (cmeiklejohn) [inline src/partisan_static_peer_service_manager.erl]: What's the motivation for removing the pending connections?  This is so things get eventually retried.

File: src/partisan_peer_service_client.erl [hunk 2]
```diff
@@ -61,8 +61,8 @@ init([Peer, From]) ->
     case connect(Peer) of
         {ok, Socket} ->
             {ok, #state{from=From, socket=Socket, peer=Peer}};
-        _Error ->
-            {stop, normal}
+        {error, Reason} ->
```
Comment (cmeiklejohn) [inline src/partisan_peer_service_client.erl]: I'm not convinced about this change - why was this necessary?  If a normal stop is no longer returned, then a client that has a closed port can trigger a client that is connected to it to crash it's entire peer service manager.

**Comment by vitorenesduarte** [comment]
I would like to know if when the join is attempted, the other node was already connected or not. But maybe this is not the best way to do it.

File: src/partisan_peer_service_client.erl (hunk 2 above)
Comment (cmeiklejohn) [inline src/partisan_peer_service_client.erl]: I'm pretty sure this change is going to prevent the simulation scenarios from working, because sometimes it takes several connection attempts before the port is opened.  Was this necessary?

**Comment by vitorenesduarte** [comment]
I also noticed that in HyParView the join is handled asynchronously, returning always ok.
Don't know if it's possible to also change that.

**Comment by cmeiklejohn** [comment]
I think modifying the maybe_connect to return additional information is fine, but I'm curious if we can make just that change in isolation without modifying the other code.

**Comment by cmeiklejohn** [comment]
Yes, the join is asynchronous because it has to wait for a response from the random walk.

**Comment by vitorenesduarte** [comment]
Basically I would like to have the result of gen_tcp:connect as the result of the join.
If not possible, I'll find another way.
